packages/flutter/lib/src/services/asset_resolver.dart

```diff
@@ -0,0 +1,88 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+import 'dart:collection';
+
+import 'asset_bundle.dart';
+
+/// Base class for resolver components that transform an input into part of
+/// a resolved asset's key.
+abstract class Qualifier {
+  String resolve();
+}
+
+/// Qualifier that knows how to find the nearest match among a list of
+/// candidates, based on a mapping to candidates from their normative values.
+/// The mapping should be sorted, e.g. by passing a [SplayTreeMap].
+///
+/// The resolution is done via linear search based on the assumption that the
+/// number of candidates is short.
+class BreakpointQualifier extends Qualifier {
+  BreakpointQualifier({ this.candidates, this.value }) {
+    assert(this.candidates.length > 0);
+  }
+  final Map<double, String> candidates;
+  final double value;
+  String resolve() {
+    double lastKey = null;
+    for (double currentKey in candidates.keys) {
+      if (lastKey != null) {
+        assert(currentKey > lastKey);
```

Comment:
> which rules out HashMap

Right, it just didn't rule out a sorted LinkedHashMap, which is convenient if you're using a literal you know is already sorted.  But you can just convert the literal to a SplayTreeMap and store it statically, which is what I'm about to do.